### ISSUE TITLE
Интеграция с VK

### ISSUE THREAD

**Issue by kulakovt** [DotNetRu/Server#1]
Необходимо найти или разработать компонент, который предоставит возможность взаимодействовать с соц. сетью [VKontakte](https://vk.com/).

Обязательные требования:

- [ ] не интерактивное выполнение всех действий;
- [ ] работа с ключами доступа, без использования пароля пользователя (OAuth и т.д.);
- [ ] работа от имени группы;
- [ ] возможность постить записи в группы (и публичные страницы). Посты должны поддерживать мета-разметку (OG) ссылок и вложения в виде картинки (как стандартный клиент);
- [ ] возможность репостить чужие записи;
- [ ] получение списка постов в группе;
- [ ] получение комментариев к постам, фотографиям и всему остальному где есть коментарии;
- [ ] получение количества просмотров, лайков и репостов для конкретной записи.

**Comment by kulakovt** [comment]
Будет удобно сделать единое хранилище защищённых данных (API-ключи, пароли, пользователи, Токены). И универсальный контракт для доступа к нему из разных сервисов. [Интеграций будет много](https://github.com/DotNetRu/Server/labels/integration), и каждой из них обязательно необходимы ключи. Ключи могут быть у каждого сообщества свои (VK, Twitter, Telegram), так и общии между всеми (YouTube, SpeakerDesk).

@Sergey-Buyanov , как считаешь?

**Comment by Sergey-Buyanov** [comment]
Да, хранить где-то это всё надо, как вариант в нашей базе, при запросе будет происходить аутентификация + авторизация в том числе на нашей стороне. Я попросил @V0v1kkk сделать research по интеграции с vk, исходя их этого уже будем реализовывать.

**Comment by V0v1kkk** [comment]
Для .Net'а самой адекватной библиотекой является [VkNet](https://vknet.github.io/vk/). С ней и смотрел.

Отвечаю по пунктам:

## не интерактивное выполнение всех действий

Тут всё ок.

## работа с ключами доступа, без использования пароля пользователя (OAuth и т.д.)

Так завести не удалось. Для аутентификации в ВК нужно передать **креды юзера + ApplicationId**. У ВК аутентификация только [интерактивная или password](https://vk.com/dev/access_token) (библиотека _VkNet_ просто эмулирует пользователя). Т.е. админу группы (или той учётке, у которой есть права писать в группу) нужно **зарегистрировать приложение, отключить 2х факторную аутентификацию, указать приложению креды, AppId**.

> API групп у ВК находиться в зачаточном состоянии. Через него [мало что можно делать](https://vk.com/dev/methods). В основном там функционал по общению юзер - группа.

## возможность постить записи в группы (и публичные страницы). Посты должны поддерживать мета-разметку (OG) ссылок и вложения в виде картинки (как стандартный клиент)

С возможность читать и писать всё ок. Всё это происходит с правами юзера. Всё что доступно ему - будет доступно приложению. 

**OG-разметку (и вообще любую другую) использовать напрямую в постах нельзя.**

### OG-разметка

OG-разметка помещается в хеадер страницы, на которую добавляется ссылка в пост. ВК сам стягивает содержимое тегов на странице, и включает в представление поста.

Например платформа [telegra.ph](http://telegra.ph) автоматически генерирует метоописание для своих статей: **[например](http://telegra.ph/Mikrodozing-03-13) (смотреть в head)**. Т.е. мы можем или публиковать какой-то контент у себя, и ссылаться на него, или найти готовую платформу для публикаций.

### Wiki

Альтернативный вариант с разметкой - создавать wiki-страницы, которые можно аттачить к постам. Но выглядит это не очень, на мой взгляд.

## возможность репостить чужие записи

Без проблем.

## получение списка постов в группе

Ок. Вычитывать можно пачками максимум по 100 записей.

## получение комментариев к постам, фотографиям и всему остальному где есть коментарии

Без проблем, но постоянно перечитывать не хочется. Нужно посмотреть в сторону _сallback API_ ([раз](https://vk.com/dev/callback_api), [два](https://vk.com/dev/using_longpoll)), или _[long poll'а](https://vk.com/dev/using_longpoll)_. Пока не смотрел.

## получение количества просмотров, лайков и репостов для конкретной записи

Всё то же самое, что по предыдущему.

## Реализация

Мне нужны входные/выходные модели. Ребята, можете задизайнить?

## Хранение аутентификационных данных

Кто-нибудь работал с такого рода специализированными репозиториями?

**Comment by kulakovt** [comment]
> Для аутентификации в ВК нужно передать креды юзера + ApplicationId

По твоей ссылки есть 2 способа, которые нам подходят:

1. Authorization code flow
1. "выбрать вкладку «Работа с API» и нажать «Создать ключ доступа»."

Также там написано о полной поддержке OAuth 2.0. Цель которого как раз делегирование прав без необходимости передавать пароли.

У нас нет цели "не говорить" пароль. Цель - не хранить его на сервере. Т.е. схема должна быть примерно такая:

1. я, как администратор, захожу на наш Server и создаю привязку Сообщества с VK группой
1. Server должен перенаправить меня на сайт VK, на котором я введу свой пароль и разрешу Server'у управлять группой
1. за это VK выдаст Server'у Token, который тот должен сохранить и использовать всякий раз для осуществления операций от моего имени
1. скорее всего ещё нужен будет refresh Token
1. Т.о. получается, что я не сообщаю серверу свой пароль (я сообщаю его VK), и на сервере мой пароль нигде не хранится (хранится только Token).

> OG-разметку (и вообще любую другую) использовать напрямую в постах нельзя.

Я немного не точно написал. Мы не хотим сами форматировать OG. Мы хотим убедиться, что если запостать ссылку с помощью API, то VK подхватит и отобразит её OG не хуже того, как он делает для ручного поста.

>  Нужно посмотреть в сторону сallback API или long poll'а.

Для этого нужно дождаться когда Сервер оживёт.

> Мне нужны входные/выходные модели. Ребята, можете задизайнить?

Что конкретно нужно?
@Sergey-Buyanov мы уже влили Мерсоновские модели? Какой там у них статус, можно использовать?

> Кто-нибудь работал с такого рода специализированными репозиториями?

Я работал. Но для тебя, как для потребителя должно быть всё прозрачно. Ты просто попросишь в зависимостях все секреты для VK и тебе из отдадут. Вся сложность по хранению и защите ложиться на плечи провайдера.

**Comment by Sergey-Buyanov** [comment]
@V0v1kkk  в VKNet  всё есть для того, что описал @AnatolyKulakov или что-то надо писать самим?
Также можешь сделать sample по тому что писал Анатолий
-  Authorization code flow;
-  работа от имени группы;
- возможность постить записи в группы (и публичные страницы). Посты должны поддерживать 
    мета-разметку (OG) ссылок и вложения в виде картинки (как стандартный клиент);
Например, если запостать ссылку с помощью API, то VK подхватит и отобразит её OG
- возможность репостить чужие записи;
- получение списка постов в группе;
- получение комментариев к постам, фотографиям и всему остальному где есть коментарии;
- получение количества просмотров, лайков и репостов для конкретной записи.

Если в чём то трудности, сложности скажи, обсудим. По authorization code flow, также могу посмотреть, что-то делали, но без vknet

**Comment by Sergey-Buyanov** [comment]
Ещё сейчас нам в первую очередь нужна аутентификация

**Comment by kulakovt** [comment]
Думаю, на первом этапе можно сократить до:

- аутентификация
- возможность постить записи в группы

Это уже позволит сделать примитивный сценарий Публикации сообщения о новом митапе. Всё остальное докрутим позже.

@V0v1kkk , для аунтефикации нужно всё-таки сделать так чтобы сервер не хранил пароль. Разобраться как это сделать в VK. Мне кажется, я видел приложения, которые позволяют так работать. Ну и Auth 2 тоже намекает.
@Sergey-Buyanov , для публикации надо начать с модели. Нам нужнен Домен Митапов для формирования по нему текста анонса.

**Comment by Sergey-Buyanov** [comment]
Вообщем план такой, как я вижу
1) нужно сделать компонент для работы c vk, с двумя фичами
- аутентификация через Authorization code flow;
- возможность постить записи в группы
интерфейс можно согласовать письменно или созвониться. Но сначала нужен sample, чтобы понять, что там должно быть. @V0v1kkk можешь сделать? По первому, если сложности я помогу
2) Подготовить домен митапов и его интегрируем с компонентом.

**Comment by egorikas** [comment]
Посмотрел я по вк-ключам и ходу авторизации, мы на хакатоне немного по-другому делали, поэтому сюда не подойдет. Но я помню, что получали 1 раз токен с[ offline ](https://vk.com/dev/permissions) и не парились по поводу его возобновления

**Comment by egorikas** [comment]
+ пароль не нужно хранить, там нормальный auth2 с [переходом на сайт вк-шечки и авторизационным кодом](https://vk.com/dev/authcode_flow_user?f=1.%20%D0%9E%D1%82%D0%BA%D1%80%D1%8B%D1%82%D0%B8%D0%B5%20%D0%B4%D0%B8%D0%B0%D0%BB%D0%BE%D0%B3%D0%B0%20%D0%B0%D0%B2%D1%82%D0%BE%D1%80%D0%B8%D0%B7%D0%B0%D1%86%D0%B8%D0%B8)

**Comment by kulakovt** [comment]
Может будет полезно
https://habr.com/post/431090/

**Comment by kulakovt** [comment]
Ещё полезная статья. У автора их несколько про VK. Также у него есть опыт, можно связаться для вопросов.
https://habr.com/ru/post/438874/
https://habr.com/ru/post/441720/
https://habr.com/ru/post/492102/

Поддержка библиотеки:
https://t.me/VkDotNet
